pkg/forwarder/forwarder.go

```diff
@@ -40,91 +48,249 @@ func init() {
+		if err != nil {
+			return nil, fmt.Errorf("failed to read from-ca-file: %v", err)
+		}
+		if !pool.AppendCertsFromPEM(data) {
+			log.Printf("warning: no certs found in from-ca-file")
+		}
+		fromTransport.TLSClientConfig.RootCAs = pool
+	}
+	fromClient := &http.Client{Transport: fromTransport}
+	if cfg.Debug {
+		fromClient.Transport = telemeterhttp.NewDebugRoundTripper(fromClient.Transport)
+	}
+	if len(cfg.FromToken) == 0 && len(cfg.FromTokenFile) > 0 {
+		data, err := ioutil.ReadFile(cfg.FromTokenFile)
+		if err != nil {
+			return nil, fmt.Errorf("unable to read from-token-file: %v", err)
+		}
+		cfg.FromToken = strings.TrimSpace(string(data))
+	}
+	if len(cfg.FromToken) > 0 {
+		fromClient.Transport = telemeterhttp.NewBearerRoundTripper(cfg.FromToken, fromClient.Transport)
+	}
+	w.fromClient = metricsclient.New(fromClient, cfg.LimitBytes, w.interval, "federate_from")
+
+	// Create the `toClient`.
+	toClient := &http.Client{Transport: metricsclient.DefaultTransport()}
+	if cfg.Debug {
+		toClient.Transport = telemeterhttp.NewDebugRoundTripper(toClient.Transport)
+	}
+	if len(cfg.ToToken) == 0 && len(cfg.ToTokenFile) > 0 {
+		data, err := ioutil.ReadFile(cfg.ToTokenFile)
+		if err != nil {
+			return nil, fmt.Errorf("unable to read to-token-file: %v", err)
+		}
+		cfg.ToToken = strings.TrimSpace(string(data))
+	}
+	if (len(cfg.ToToken) > 0) != (cfg.ToAuthorize != nil) {
+		return nil, errors.New("an authorization URL and authorization token must both specified or empty")
+	}
+	if len(cfg.ToToken) > 0 {
+		// Exchange our token for a token from the authorize endpoint, which also gives us a
+		// set of expected labels we must include.
+		rt := authorize.NewServerRotatingRoundTripper(cfg.ToToken, cfg.ToAuthorize, toClient.Transport)
+		toClient.Transport = rt
+		transformer.WithFunc(func() metricfamily.Transformer {
+			return metricfamily.NewLabel(nil, rt)
+		})
+	}
+	w.toClient = metricsclient.New(toClient, cfg.LimitBytes, w.interval, "federate_to")
+	w.transformer = transformer
 
-		transformer: t,
-		rules:       rules,
+	// Configure the matching rules.
+	var rules []string
+	if len(cfg.RulesFile) > 0 {
+		data, err := ioutil.ReadFile(cfg.RulesFile)
+		if err != nil {
+			return nil, fmt.Errorf("unable to read match-file: %v", err)
+		}
+		rules = append(cfg.Rules, strings.Split(string(data), "\n")...)
+	}
+	for i := 0; i < len(rules); {
+		s := strings.TrimSpace(rules[i])
+		if len(s) == 0 {
+			rules = append(rules[:i], rules[i+1:]...)
+			continue
+		}
+		rules[i] = s
+		i++
 	}
+	w.rules = rules
+
+	return &w, nil
 }
 
-func (w *Worker) LastMetrics() []*clientmodel.MetricFamily {
+// Reconfigure temporarily stops a worker and reconfigures is with the provided Config.
+// Is thread safe and can run concurrently with `LastMetrics` and `Run`.
+func (w *Worker) Reconfigure(cfg Config) error {
+	worker, err := New(cfg)
+	if err != nil {
+		return fmt.Errorf("failed to reconfigure: %v", err)
+	}
+
 	w.lock.Lock()
```

Comment:
I decided against using the channel to pass configuration to the worker. Instead, I chose to lock here, keep all the reconfiguration inside the `Reconfigure` func, and signal the `Run` func. Otherwise, if we instead passed a Config or Worker down the channel to the `Run` func, the `Reconfigure` func would not fully encapsulate a reconfiguration, meaning that any process that reconfigured a worker and then made a request against it, would be racing against the `Run` func. This keeps things deterministic and testable.